### ISSUE TITLE
The application is automatically made available under machine’s detected IP/hostname

### ISSUE THREAD

**Issue by kwkwan** [ietf-ribose/bibxml-service#122]
When I relaunch the web pod, I got error 400 Bad Request.  The logs are look like:

```
Invalid HTTP_HOST header: '10.0.1.85:8000'. You may need to add '10.0.1.85' to ALLOWED_HOSTS.
Bad Request: /
```

As we are using fargate and the IP address will be changed every time it relaunch.  I manually update the code to add the ip address into `ALLOWED_HOSTS` for a dirty fix.  It is better to detect the ip address and add it into `ALLOWED_HOSTS`.

After browsing the internet, the following listed code can add the ip address dynamically.

```
from socket import gethostname, gethostbyname, gethostbyname_ex
ALLOWED_HOSTS = [ gethostname(), ] + list(set(gethostbyname_ex(gethostname())[2]))
```

Please update the code to allow adding ip address dynamically.  Thanks!

**Comment by strogonoff** [comment]
AWS CF docs have an article on Host header: https://aws.amazon.com/premiumsupport/knowledge-center/configure-cloudfront-to-forward-headers/

**Comment by kwkwan** [comment]
This situation happens during the health check stage by ALB of the container, not during forwarding the request.

When the container initialises, Django will check the `ALLOWED_HOSTS`.  As the container is created by fargate and a random private ip will be assigned to the container at this stage.  As the automatically assigned ip is not in the `ALLOWED_HOSTS`, error 400 will be retuned for the health check request.  The health check request is created by the ALB to determine the pod is running and attached the container to the ALB if the container is healthy.

If ALB cannot find a healthy container (target), it will turn 503 error.

By the way, we do not use cloudfront in AWS and X-Forward headers will be created by ALB automatically.

**Comment by strogonoff** [comment]
> When the container initialises, Django will check the `ALLOWED_HOSTS`. As the container is created by fargate and a random private ip will be assigned to the container at this stage. As the automatically assigned ip is not in the `ALLOWED_HOSTS`, error 400 will be retuned for the health check request. The health check request is created by the ALB to determine the pod is running and attached the container to the ALB if the container is healthy.

@kwkwan  Hmm OK, I see that we must add a hostname for healthcheck requests? We can add an additional internal domain name to allowed hosts using `INTERNAL_HOSTNAMES` environment variable, it can be a comma-separated list of extra hostnames Django should allow: https://github.com/ietf-ribose/bibxml-service/blob/main/bibxml/settings.py#L57-L70

**Comment by strogonoff** [comment]
1. I don’t believe this should be part of concern to application level.
2. This makes the service always accessible by its IP, unless there is extra configuration in the front of it. You are tailoring application to K8s, but what’s being delivered here is not K8s config but this repository with containers and Compose configuration.

I’ll see how this can be resolved in some other way. Ideally the infrastructure should know its IPs and not require the app to inspect interfaces. Or there could be a special configuration option.

**Comment by ronaldtse** [comment]
> I see that we must add a hostname for healthcheck requests?

Unfortunately AWS does not allow that -- there will not be a hostname given. It will be some IP address from the AWS load balancer, which uses a rather random private IP address. If we *really* want to use the private IP address of the load balancer, since it is not available through static infrastructure (at time of deployment, through ENV vars), Django must read it in dynamically somehow with a pre-start request.

> This makes the service always accessible by its IP, unless there is extra configuration in the front of it.

Well, in AWS Lambda/Fargate right now, they are behind private IPs within a private subnet. So there is no concern with security here.

> Ideally the infrastructure should know its IPs and not require the app to inspect interfaces. Or there could be a special configuration option/environment variable.

The bottom line is this is not supported in such AWS cloud setup. The infrastructure uses dynamic IPs and they change frequently. And these values are not available in environment variables (remember that the ENV vars are statically set).

It is however possible to have a "range" of private IP address of the private subnet. That can be available as static information. Would you prefer that?

**Comment by strogonoff** [comment]
> Unfortunately AWS does not allow that -- there will not be a hostname given. It will be some IP address from the AWS load balancer, which uses a rather random private IP address. If we really want to use the private IP address of the load balancer, since it is not available through static infrastructure (at time of deployment, through ENV vars), Django must read it in dynamically somehow with a pre-start request.

I see.

> The bottom line is this is not supported in such AWS cloud setup.

OK, I see.

> That can be available as static information. Would you prefer that?

Django doesn’t support ranges in ALLOWED_HOSTS. We could do it with a custom middleware, but what I can do instead is to introduce an environment variable that will explicitly tell Django “accept requests that use machine’s internal IP, because we will ensure they only come from internal healthcheck”.

**Comment by ronaldtse** [comment]
Aha, thanks @strogonoff !

**Comment by strogonoff** [comment]
Should

**Comment by ronaldtse** [comment]
> Please set one or both of them to 1 in bibxml-infrastructure as desired to enable auto IP/hostname detection.

@kwkwan can you help with this?

**Comment by kwkwan** [comment]
Auto IP/hostname detection has been enabled.  Refer to https://github.com/ietf-ribose/bibxml-infrastructure/issues/23

**Comment by strogonoff** [comment]
Okay, this can be closed I think.

**Comment by ronaldtse** [comment]
Thanks @strogonoff !
